### PR TITLE
Tracelocation qrcode

### PR DESCRIPTION
**Pull request by svengabr**
Wishlist: https://github.com/corona-warn-app/cwa-wishlist/issues/399
Continue of: https://github.com/corona-warn-app/cwa-website/pull/1130

Some minor changes to the original work of @bastianjoel 

Still a work in progress as we are missing a small disclaimer that still needs to be finalized.

### CODE REVIEW

File: src/data/eventregistration_de.json [hunk 1]
```diff
@@ -3,6 +3,8 @@
         "headline": {
             "title": "QR-Code für Event Check-In erstellen"
         },
+        "intro": "Die CWA ermöglicht nun einen Check-In für Events und Orte. Personen, die Events veranstalten oder ein Geschäft haben, können nun einen QR-Code erstellen. Durch Scannen des QR-Codes können sich Gäste bei Ankunft einchecken, um si ihre Anwesenheit zu registrieren. Auf Wunsch legt die App außerdem einen entsprechenden tagebuch-Eintrag an. Wird eine eingecheckte Person später positiv auf das Coronavirus getestet, können andere Personen gewarnt werden, die zur selben Zeit eingecheckt waren.",
```
Comment (Ein-Tim) [inline src/data/eventregistration_de.json:6]: "[…] um si ihre Anwesenheit […]" -> "[…] um so ihre Anwesenheit […]"
"[…] entsprechenden tagebuch-Eintrag an […]" -> […] entsprechenden Tagebuch-Eintrag an […]

File: src/data/eventregistration.json [hunk 2]
```diff
@@ -3,6 +3,8 @@
         "headline": {
             "title": "Create QR code for event check-in"
         },
+        "intro": "The CWA now allows check-in for events and venues. People who host events or have a business can now create a QR code. By scanning the QR code, guests can check in upon arrival to register si their attendance. If desired, the app will also create a corresponding diary entry. If a checked-in person later tests positive for the coronavirus, other people who were checked in at the same time can be warned.",
```
Comment (Ein-Tim) [inline src/data/eventregistration.json:6]: "[…] to register si their […]" -> "[…] to register their […]"

File: src/data/eventregistration.json [hunk 3]
```diff
@@ -3,6 +3,8 @@
...
+        "disclaimer": "All information is processed only locally in your browser. Neither the QR codes you generate nor the information they contain are sent to the Corona Warning app servers. The source code of the generator is freely available for viewing.",
```
Comment (Ein-Tim) [inline src/data/eventregistration.json:7]: "Corona Warning app servers" -> "Corona-Warn-App servers"

File: src/pages/en/eventregistration/index.html [hunk 4]
```diff
@@ -0,0 +1,8 @@
+---
+lang_en: true
+page-title: Open-Source Project Corona-Warn-App - Event QR Code
+page-description:
```
Comment (bastianjoel) [inline src/pages/en/eventregistration/index.html:4]: @svengabr I've missed to add the page description in English. 

For the German description I used the text above the "create qr code" button on the start page in the android app.

**Review by Ein-Tim — CHANGES_REQUESTED** [review]
Some comments on the texts. 🙂

**Review by bastianjoel — CHANGES_REQUESTED** [review]

**Review by Ein-Tim — COMMENTED** [review]
Some more things:

1. The checkout after minutes should be restricted to 15 min blocks. The app only accepts those, if you enter f.e. 61 minutes into the generator the app will show 1h 15 min. - **RESOLVED**
2. The red stars should be explained somewhere.
3. The download now is a PNG, is this intended? - **RESOLVED** 
4. The QR-Code disappears from the website when the button "Download" is clicked. - **RESOLVED**

File: src/data/eventregistration.json (hunk 3 above)
Comment (svengabr) [reply src/data/eventregistration.json]: Done, thanks!

File: src/pages/en/eventregistration/index.html (hunk 4 above)
Comment (svengabr) [reply src/pages/en/eventregistration/index.html]: I have added the english texts, thanks!

File: src/data/eventregistration.json (hunk 2 above)
Comment (svengabr) [reply src/data/eventregistration.json]: Done!

File: src/data/eventregistration_de.json (hunk 1 above)
Comment (svengabr) [reply src/data/eventregistration_de.json]: Done

Comment (dsarkar) [reply src/data/eventregistration_de.json]: Die CWA --> Die Corona-Warn-App

File: src/data/eventregistration.json (hunk 2 above)
Comment (dsarkar) [reply src/data/eventregistration.json]: The Corona-Warn-App

**Comment by svengabr** [comment]
* Fixed all the typos.
* The original code has a generation of PNG, this is intended I, guess.
* The QR code disappearing was fixed

**Comment by bastianjoel** [comment]
@Ein-Tim

1. The input field has the step attribute set to "15"  but some browser don't validate that. Because the qr code still works with an "invalid" time I would say that the current solution is sufficient. 

3. Yes, this is intended. Exporting as pdf would require another library. I considered that but had the following problems.   
If this is still wanted we would need to use a library like pdfmake, a library wich convertes canvas to pdf or pdf-lib. 

The problem with pdfmake and generation from canvas is, that we can't add an pdf as background and need to use an image background. Because of that we wouldn't have any benefits in resolution and size compared to a pdf (this isn't the case if somebody implements the pdf as pdfmake template). 

With pdf-lib on the other hand I had problems including it into the project and using it in a clean way, because it wasn't possible to use "async" or "await".

**Comment by Art4** [comment]
I've noted that a date in the format `5.5.21` is not valid. For better UX we coud try to automatically format the date if the field lost the focus.
From `5.5.21` to `05.05.2021`
If the date is invalid/could not be formated the value should not be changed.

**Comment by Ein-Tim** [comment]
@bastianjoel 

1 - I agree with you! 

3 - Okay, it'a also nice how it is! 

Again thank you so much! ❤️

**Comment by dsarkar** [comment]
@Art4 Thanks for pointing that out. We understand that in the following days and weeks there will be many suggestions on how to improve UX and implement enhanced features. We will take this into account.

**Comment by Ein-Tim** [comment]
@svengabr 
Although the QR-Code is not disappearing anymore now, it now changes on the website when you click "Download". This could result that the person who created the QR-Code checks in on the website after he/she downloaded the QR-Code and thus he/she checks into another event. They will have the same name and everything will be the same, but it's not the same event and so he/she won't be warned. 
This should be fixed IHMO.

**Comment by fynngodau** [comment]
Possibly, it would even be better if clicking Show does not generate a new code each time, at least not without warning, as users might not be aware that two similar codes would actually be two different events.

**Comment by dsarkar** [comment]
FYI https://www.coronawarn.app/en/blog/2021-04-23-qr-code-generator/

File: src/data/eventregistration_de.json [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+                "description": "z.B. Einkaufszentrum",
+                "address": "z.B. Musterstr. 12, 12345 Musterstadt",
+                "defaultcheckinlength": "z.b. 45"
+            },
+            "errors": {
+                "descriptionrequired": "Es wird eine Beschreibung benötigt",
+                "descriptionmax": "Die Beschreibung darf maximal 100 Zeichen lang sein",
+                "addressrequired": "Es wird eine Adresse benötigt",
+                "addressmax": "Die Adresse darf maximal 100 Zeichen lang sein",
+                "defaultcheckinlengthrequired": "Bitte geben Sie eine typische Aufentshaltsdauer an",
```
Comment (DerVogel2020) [inline src/data/eventregistration_de.json:29]: Aufenthaltsdauer

File: src/data/eventregistration_de.json [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+            "title": "QR-Code für Event Check-In erstellen"
+        },
+        "intro": "Die App ermöglicht nun einen Check-In für Events und Orte. Personen, die Events veranstalten oder ein Geschäft haben, können nun einen QR-Code erstellen. Durch Scannen des QR-Codes können sich Gäste bei Ankunft einchecken, um so ihre Anwesenheit zu registrieren. Auf Wunsch legt die App außerdem einen entsprechenden Tagebuch-Eintrag an. Wird eine eingecheckte Person später positiv auf das Coronavirus getestet, können andere Personen gewarnt werden, die zur selben Zeit eingecheckt waren.",
+        "disclaimer": "Alle Informationen werden ausschließlich lokal in Ihrem Browser verarbeitet. Weder die von Ihnen generierten QR-Codes noch die darin enthaltene Informationen werden an die Server der Corona-Warn-App gesendet. Der Quelltext des Generators ist frei einsehbar.",
+        "form": {
+            "fields": {
+                "locationtype": "Ort/Event",
+                "description": "Beschreibung",
+                "address": "Anschrift",
+                "defaultcheckinlength": "Typische Aufentshaltsdauer (Minuten)",
```
Comment (DerVogel2020) [inline src/data/eventregistration_de.json:13]: Aufenthaltsdauer

File: src/data/eventregistration_de.json [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+                "starttime": "Startzeit",
+                "endtime": "Endzeit",
+                "show": "Erstellen",
+                "download": "Herunterladen"
+            },
+            "placeholders": {
+                "description": "z.B. Einkaufszentrum",
+                "address": "z.B. Musterstr. 12, 12345 Musterstadt",
+                "defaultcheckinlength": "z.b. 45"
```
Comment (DerVogel2020) [inline src/data/eventregistration_de.json:22]: z. B.

**Review by DerVogel2020 — COMMENTED** [review]
Just found 3 typos.

**Comment by Ein-Tim** [comment]
@DerVogel2020 

Do you want to provide a PR for this? If not I'll quickly do it.

**Comment by DerVogel2020** [comment]
@Ein-Tim please do it.

**Comment by Ein-Tim** [comment]
@DerVogel2020 See #1135.

**Comment by heinezen** [comment]
> @svengabr
> 
> Although the QR-Code is not disappearing anymore now, it now changes on the website when you click "Download". This could result in a case where the person who created the QR-Code checks in to the QR-Code shown on the website after he/she downloaded the QR-Code and thus he/she checks in to another event than the people who check in with the QR-Code which was downlaoded.
> They will have the same name and everything, but it's not the same event and so he/she won't be warned in case somebody later tests positive.
> This should be fixed IHMO.

Can you make an issue for this so it doesn't get lost?

**Comment by Ein-Tim** [comment]
@heinezen 

I'll create some more issues reg. the generator today (:

**Pull request merged**